src/util/provenanceUtils.js

```diff
@@ -0,0 +1,150 @@
+const { resolveSchema, ServerError } = require('@projecttacoma/node-fhir-server-core');
+const { v4: uuidv4 } = require('uuid');
+
+/**
+ * Creates a raw JSON AuditEvent resource from the X-Provenance headers of a submission request
+ * @param {string} provenance the provenance headers in string form from the request
+ * @param {Object} args the object containing the base_version parameter
+ * @returns A JSON object representing an AuditEvent to be stored in the system
+ */
+const createAuditEventFromProvenance = (provenance, args) => {
+  provenance = JSON.parse(provenance);
+  const audit = {};
+  audit.type = {
+    system: 'http://dicom.nema.org/resources/ontology/DCM',
+    code: '110100',
+    display: 'Application Activity'
+  };
+  audit.period = provenance.occurredPeriod;
+  audit.recorded = provenance.recorded;
+  // 0 represents a successful request
+  // TODO: Add functionality for catching an unsuccessful request
+  audit.outcome = 0;
+  audit.purposeOfEvent = [];
+  if (provenance.reason) {
+    audit.purposeOfEvent.push(...provenance.reason);
+  }
+
+  switch (provenance?.activity?.coding?.[0]?.code) {
+    case 'UPDATE':
+      audit.action = 'U';
+      break;
+    default:
+      // Default to C for CREATE for now
+      audit.action = 'C';
+      break;
+  }
+  let source = {};
+  audit.agent = [];
+  if (provenance.agent) {
+    provenance.agent.forEach(agent => {
+      agent['requestor'] = true;
+
+      if (agent.role.coding.filter(role => role.code === 'AGNT').length > 0) {
```

Comment:
```suggestion
      if (agent.role.coding.some(role => role.code === 'AGNT')) {
```

nitpick: this function does the same thing but intuitively returns a boolean rather than checking the length of a filtered array

Reply:
So useful. Didn't know about this one. Thanks!